### DISCUSSION TITLE
Error with sending account verification email with named auth class

### DISCUSSION THREAD

**Discussion by Samuelodan** [janko/rodauth-rails#232] [Q&A] [answered]
Hi there,
I get this error when I try to send out emails after sign up:

```
[ActiveJob] [ActionMailer::MailDeliveryJob] [fbd36e5a-7fa2-4488-b745-1d641fa6a097]   ↳ app/mailers/rodauth_mailer.rb:58:in `block in rodauth'
[ActiveJob] [ActionMailer::MailDeliveryJob] [fbd36e5a-7fa2-4488-b745-1d641fa6a097] RodauthMailer#verify_account: processed outbound mail in 7.1ms
[ActiveJob] [ActionMailer::MailDeliveryJob] [fbd36e5a-7fa2-4488-b745-1d641fa6a097] Error performing ActionMailer::MailDeliveryJob (Job ID: fbd36e5a-7fa2-4488-b745-1d641fa6a097) from Async(default) in 12.1ms: ArgumentError (unknown rodauth configuration: nil):
/Users/samuelodan/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rodauth-rails-1.11.0/lib/rodauth/rails/app.rb:62:in `rodauth!'
/Users/samuelodan/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rodauth-rails-1.11.0/lib/rodauth/rails.rb:57:in `model'
```

This started happening after I versioned the auth class, :v1.

Here are some details.
`:v1` is the name passed into the Mailer action and it represents the `V1::RodauthMain` auth class.

Before the versioning, `nil` was the argument passed into the Mailer action's `name` parameter and that worked fine. It represented the `RodauthMain` auth class (without a name).

Please let me know if you need any more info.
Thanks.

Edit: I did a bit more testing and found that it isn't because the auth class is versioned. It just fails if I name the auth class at all. Even `rodauth(:admin)` registered as `configure RodauthMain, :admin` and placed in `app/misc/rodauth_main.rb` will fail.

**Comment by janko** [comment]
Yes, it seems your mailer configuration relies on a default Rodauth configuration existing (without a name), which doesn't exist in your setup. Could you post the part of your Rodauth configuration that calls the mailer, as well as the mailer code?

**Reply by Samuelodan** [reply]
Thanks for your swift response, Janko. Here is the code:

```rb
# rodauth_main.rb
    create_verify_account_email do
      RodauthMailer.verify_account(self.class.configuration_name, account_id, verify_account_key_value)
    end
    
# rodauth_mailer.rb
  def verify_account(name, account_id, key)
    @rodauth = rodauth(name, account_id) { @verify_account_key_value = key }
    @account = @rodauth.rails_account

    mail subject: @rodauth.email_subject_prefix + @rodauth.verify_account_email_subject
  end

  private

  def rodauth(name, account_id, &block)
    instance = RodauthApp.rodauth(name).allocate
    instance.instance_eval { @account = account_ds(account_id).first! }
    instance.instance_eval(&block) if block
    instance
  end    
```

**Reply by janko** [reply]
Hmm, this code is identical to the offical demo app, and mailer integration works there, so there must be more at play here. Could you post your Rodauth app code as well?

**Reply by Samuelodan** [reply]
Okay, here it is:
```rb
class RodauthApp < Rodauth::Rails::App
  # primary configuration
  configure V1::RodauthMain, :v1

  route do |r|
    r.rodauth(:v1) # route rodauth requests
  end
end
```

**Reply by janko** [reply, marked as answer]
I only now took a look at the backtrace 🤦🏻‍♂️ You'll notice it's not related to emails at all, but to the `Rodauth::Rails.model` call in your account model file, you need to change it to `Rodauth::Rails.model(:v1)` (or to `Rodauth::Model(V1::RodauthMain)`, which new rodauth-rails versions will generate).

**Reply by Samuelodan** [reply]
Ah, it works now.
> You'll notice it's not related to emails at all

Oh, that makes sense, I didn't understand how it worked so it wasn't obvious. Adding even two more lines to the backtrace like so:
```
[ActiveJob] [ActionMailer::MailDeliveryJob] [4a55420b-3837-4057-8529-f9a65d846139]   ↳ app/mailers/rodauth_mailer.rb:58:in `block in rodauth'
[ActiveJob] [ActionMailer::MailDeliveryJob] [4a55420b-3837-4057-8529-f9a65d846139] RodauthMailer#verify_account: processed outbound mail in 5.7ms
[ActiveJob] [ActionMailer::MailDeliveryJob] [4a55420b-3837-4057-8529-f9a65d846139] Error performing ActionMailer::MailDeliveryJob (Job ID: 4a55420b-3837-4057-8529-f9a65d846139) from Async(default) in 6.88ms: ArgumentError (unknown rodauth configuration: nil):
/Users/samuelodan/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rodauth-rails-1.11.0/lib/rodauth/rails/app.rb:62:in `rodauth!'
/Users/samuelodan/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rodauth-rails-1.11.0/lib/rodauth/rails.rb:57:in `model'
/Users/samuelodan/repos/vercir/vercir-api/app/models/user.rb:2:in `<class:User>'
/Users/samuelodan/repos/vercir/vercir-api/app/models/user.rb:1:in `<main>'
```
Would have made it a little easier to spot.
Thank you so much for all the help.

**Reply by Samuelodan** [reply]
So, how do I manage multiple versions? Would it be alright to include multiple `Rodauth::Model` modules in the account model, each with their distinct names passed in?

**Reply by janko** [reply]
I think there is no need, because different versions will need to share the same table/column names, so they'll share the same associations. It should be enough to include the latest version's model mixin.

**Reply by Samuelodan** [reply]
hmm, interesting. So, if I understand you correctly, I can expect `:v1` to still work even if only `:v2` is included in the model.
Aiit then, thank you once again. All our interactions have greatly increased my confidence in Rodauth, and I hope to someday (sooner, rather than later) understand how it all works; rodauth-rails too.
I read your article on what it took to build rodauth-rails, but that was before I'd even installed it, so I'll give it another read now.

Edit: So, I tried it out and it worked just as you said.
